hermes/bin/src/event/queue.rs

```diff
@@ -42,52 +43,57 @@ pub(crate) enum Error {
     /// Failed when event queue not been initialized.
     #[error("Event queue not been initialized. Call `HermesEventQueue::init` first.")]
     NotInitialized,
-}
 
-/// Hermes event queue execution loop thread handler
-pub(crate) type HermesEventLoopHandler = JoinHandle<anyhow::Result<()>>;
+    /// Event loop has crashed unexpectedly.
+    #[error("Event loop has crashed unexpectedly.")]
+    EventLoopPanics,
+}
 
 /// Hermes event queue.
 /// It is a singleton struct.
-pub(crate) struct HermesEventQueue {
+struct HermesEventQueue {
     /// Hermes event queue sender
     sender: Sender<HermesEvent>,
 }
 
-impl HermesEventQueue {
-    /// Creates a new instance of the `HermesEventQueue`.
-    /// Runs an event loop thread.
-    ///
-    /// # Errors:
-    /// - `Error::AlreadyInitialized`
-    pub(crate) fn init(indexed_apps: Arc<IndexedApps>) -> anyhow::Result<HermesEventLoopHandler> {
-        let (sender, receiver) = std::sync::mpsc::channel();
-
-        EVENT_QUEUE_INSTANCE
-            .set(Self { sender })
-            .map_err(|_| Error::AlreadyInitialized)?;
-
-        let event_loop = thread::spawn(move || event_execution_loop(&indexed_apps, receiver));
-
-        Ok(event_loop)
-    }
+/// Hermes event queue execution loop thread handler
+pub(crate) struct HermesEventLoopHandler {
+    /// Hermes event queue execution loop thread handler
+    handle: Option<JoinHandle<anyhow::Result<()>>>,
```

Comment:
Why do we need `Option` at this place, seems it is redundant

Reply:
It needs to be consumed because it can only be used once.
We use `take()` which consumes the Option and returns the handle or None.  Which makes the function call safe, but only able to be consumed once.  `JoinHandle` is a problematic type...